File: ieee2022/python/disjoint_subsets/up_tree_order.py
import random
import time


class UpTreeOrder:

    def __init__(self, index):
        self.index = index
        self.parent = self
        self.order = 0

    def find(self):
        root = self
        while root.parent != root:
            root = root.parent

        active = self
        while active != root:
            next = active.parent
            active.parent = root
            active = next

        return root

    def find_path_split(self):
        active = self
        while active.parent != active:
            active, active.parent = active.parent, active.parent.parent
        return active

    def find_path_halving(self):
        active = self
        while active.parent != active:
            active.parent = active.parent.parent
            active = active.parent
        return active

    get_root = find

    @staticmethod
    def union(a, b):
        a_root = a.get_root()
        b_root = b.get_root()
        if a_root.order < b_root.order:
            a_root.parent = b_root
        elif b_root.order < a_root.order:
            b_root.parent = a_root
        else:
            a_root.parent = b_root
            b_root.order += 1

    @staticmethod
    def union_with_return(a, b):
        a_root = a.get_root()
        b_root = b.get_root()
        if a_root.order < b_root.order:
            a_root.parent = b_root
            return b_root
        if b_root.order < a_root.order:
            b_root.parent = a_root
            return a_root
        a_root.parent = b_root
        b_root.order += 1
        return b_root

    @staticmethod
    def in_same_set(a, b):
        return a.get_root() == b.get_root()


if __name__ == '__main__':
    trees = []
    total_time = 0
    total_time -= time.time()
    for i in range(1000000):
        tree = UpTreeOrder(i)
        trees.append(tree)
        UpTreeOrder.union(tree, random.choice(trees))
    total_time += time.time()
    print(total_time)

File: ieee2022/java/src/DisjointSubsets/UpTreeOrder.java
package DisjointSubsets;

import java.util.ArrayList;
import java.util.Random;

public class UpTreeOrder<T> {

    public T index;
    public UpTreeOrder<T> parent;
    public int order;

    public UpTreeOrder(T index) {
        this.index = index;
        this.parent = this;
        this.order = 0;
    }

    public UpTreeOrder<T> find() {
        UpTreeOrder<T> root = this;
        while (root.parent != root) {
            root = root.parent;
        }

        UpTreeOrder<T> active = this;
        UpTreeOrder<T> next;
        while (active != root) {
            next = active.parent;
            active.parent = root;
            active = next;
        }

        return root;
    }

    public UpTreeOrder<T> findPathSplit() {
        UpTreeOrder<T> active = this;
        UpTreeOrder<T> temp;
        while (active.parent != active) {
            temp = active.parent;
            active.parent = temp.parent;
            active = temp;
        }
        return active;
    }

    public UpTreeOrder<T> findPathHalving() {
        UpTreeOrder<T> active = this;
        UpTreeOrder<T> temp;
        while (active.parent != active) {
            active.parent = active.parent.parent;
            active = active.parent;
        }
        return active;
    }

    public UpTreeOrder<T> getRoot() {
        return this.find();
    }

    public static <T> void union(UpTreeOrder<T> a, UpTreeOrder<T> b) {
        UpTreeOrder<T> aRoot = a.getRoot();
        UpTreeOrder<T> bRoot = b.getRoot();
        if (aRoot.order < bRoot.order) {
            aRoot.parent = bRoot;
        } else if(bRoot.order < aRoot.order) {
            bRoot.parent = aRoot;
        } else {
            aRoot.parent = bRoot;
            bRoot.order++;
        }
    }

    public static <T> UpTreeOrder<T> unionWithReturn(UpTreeOrder<T> a, UpTreeOrder<T> b) {
        UpTreeOrder<T> aRoot = a.getRoot();
        UpTreeOrder<T> bRoot = b.getRoot();
        if (aRoot.order < bRoot.order) {
            aRoot.parent = bRoot;
            return bRoot;
        }
        if(bRoot.order < aRoot.order) {
            bRoot.parent = aRoot;
            return aRoot;
        }
        aRoot.parent = bRoot;
        bRoot.order++;
        return bRoot;
    }

    public static <T> boolean inSameSet(UpTreeOrder<T> a, UpTreeOrder<T> b) {
        return a.getRoot() == b.getRoot();
    }

    public static void main(String[] args) {
        Random random = new Random();
        ArrayList<UpTreeOrder<Integer>> trees = new ArrayList<>();
        long totalTime = - System.currentTimeMillis();
        for (int i = 0; i < 1000000; i++) {
            UpTreeOrder<Integer> tree = new UpTreeOrder<>(i);
            trees.add(tree);
            union(tree, trees.get(random.nextInt(trees.size())));
        }
        System.out.println((totalTime + System.currentTimeMillis()) / 1000.0);
    }
}
